conbench/api/compare.py

```diff
@@ -135,18 +136,36 @@ def __init__(
         threshold: Optional[float],
         threshold_z: Optional[float],
     ) -> None:
-        if (
-            baseline
-            and baseline.unit
-            and contender
-            and contender.unit
-            and baseline.unit != contender.unit
-        ):
-            raise UnmatchingUnitsError(
-                "Benchmark units do not match. Benchmark result with ID "
-                f"'{baseline.id}' has unit '{baseline.unit}' and benchmark result "
-                f"with ID '{contender.id}' has unit '{contender.unit}'."
-            )
+        # What do we know here? Is one of baseline and contender guaranteed
+        # to not be None?
+        assert any([baseline, contender])
```

Comment:
I encoded this assumption now here thinking that it does not make sense to compared two errored results.  Test suite passes. If in practice in fact we create a `BenchmarkResultComparator` for two errored results then I'd argue maybe we just don't want to do that.

What do we think @austin3dickey?

Reply:
In short: it is a good assumption that at least one of these will not be None. But the optionality of each benchmark result here has nothing to do with whether it has an error or not. 

When comparing two runs, we need to join pairs of benchmark results that make sense. We decided to do a full outer join on the "timeseries fingerprint" (i.e. case/context/hardware/repo). There might be some fingerprints where the baseline run has a fingerprint that is not present in the contender run, or vice versa. In those cases, when we construct the `BenchmarkResultComparator` object, the missing result will be None. But a full outer join cannot result in a pair of benchmark results where _both_ results are None. So this line of code is good.

To clarify, an errored benchmark result passed to either `baseline` or `contender` here will not be `None` -- it will be a full-fledged benchmark result where the `.error` attribute is not None. In most cases, that will also mean the `.unit` attribute is None (but not always; e.g. [the case](https://github.com/conbench/conbench/blob/fcb7e451c807a4c113cfcc9adc9bc4112ba5390e/conbench/tests/api/_fixtures.py#L131) where one iteration failed).

Reply:
> But a full outer join cannot result in a pair of benchmark results where both results are None. So this line of code is good.

:heart: